pkg/collector/listeners/kubelet_test.go

```diff
@@ -0,0 +1,118 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2017 Datadog, Inc.
+
+// +build kubelet
+
+package listeners
+
+import (
+	"testing"
+
+	"github.com/DataDog/datadog-agent/pkg/util/kubernetes/kubelet"
+	"github.com/stretchr/testify/assert"
+)
+
+func getMockedPod() *kubelet.Pod {
+	containerSpecs := []kubelet.ContainerSpec{
+		{
+			Name:  "foo",
+			Image: "datadoghq.com/foo:latest",
+			Ports: []kubelet.ContainerPortSpec{
+				{
+					ContainerPort: 1337,
+					HostPort:      1338,
+					Name:          "footcpport",
+					Protocol:      "TCP",
+				},
+				{
+					ContainerPort: 1339,
+					HostPort:      1340,
+					Name:          "fooudpport",
+					Protocol:      "UDP",
+				},
+			},
+		},
+		{
+			Name:  "bar",
+			Image: "datadoghq.com/bar:latest",
+			Ports: []kubelet.ContainerPortSpec{
+				{
+					ContainerPort: 1122,
+					HostPort:      1133,
+					Name:          "barport",
+					Protocol:      "TCP",
+				},
+			},
+		},
+	}
+	kubeletSpec := kubelet.Spec{
+		HostNetwork: false,
+		Hostname:    "mock",
+		NodeName:    "mockn-node",
+		Containers:  containerSpecs,
+	}
+	containerStatuses := []kubelet.ContainerStatus{
+		{
+			Name:  "foo",
+			Image: "datadoghq.com/foo:latest",
+			ID:    "docker://foorandomhash",
+		},
+		{
+			Name:  "bar",
+			Image: "datadoghq.com/bar:latest",
+			ID:    "docker://barrandomhash",
+		},
+	}
+	kubeletStatus := kubelet.Status{
+		Phase:      "Running",
+		PodIP:      "127.0.0.1",
+		HostIP:     "127.0.0.2",
+		Containers: containerStatuses,
+	}
+	return &kubelet.Pod{
+		Spec:   kubeletSpec,
+		Status: kubeletStatus,
+		Metadata: kubelet.PodMetadata{
+			Name: "mock-pod",
+		},
+	}
+}
+
+func TestProcessNewPod(t *testing.T) {
+	services := make(chan Service, 2)
+	listener := KubeletListener{
+		newService: services,
+		services:   make(map[ID]Service),
+	}
+	listener.processNewPod(getMockedPod())
+
+	service := <-services
```

Comment:
In case the service is not set, you can use a `select` around the channel read, with a default case making the test FailNow(), see https://gobyexample.com/non-blocking-channel-operations